File: 1298-reverse-substrings-between-each-pair-of-parentheses/1298-reverse-substrings-between-each-pair-of-parentheses.java
class Solution {
    public String reverseParentheses(String s) {
        Stack<Integer> match_par = new Stack<>();
        int[] match = new int[s.length()];
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') match_par.push(i);
            else if (s.charAt(i) == ')') {
                match[i] = match_par.peek();
                match[match_par.pop()] = i;
            }
        }
        String res = "";
        for (int i = 0, dir = 1; i < s.length(); i += dir) {
            if (s.charAt(i) == '(' || s.charAt(i) == ')') {
                i = match[i];
                dir = -dir;
            } else res += s.charAt(i);
        }
        return res;
    }
}

File: 1298-reverse-substrings-between-each-pair-of-parentheses/1298-reverse-substrings-between-each-pair-of-parentheses.cpp
class Solution {
public:
    string reverseParentheses(string s) {
        stack<int> match_par;
        vector<int> match(s.size());
        for(int i = 0; i < s.size(); i++) {
            if (s[i] == '(') match_par.push(i);
            else if (s[i] == ')') {
                match[i] = match_par.top();
                match[match_par.top()] = i;
                match_par.pop();
            }
        }
        string res;
        for (int i = 0, dir = 1; i < s.size(); i += dir) {
            if (s[i] == '(' || s[i] == ')') {
                i = match[i];
                dir = -dir;
            } else res += s[i];
        }
        return res;
    }
};